# Learning

This notebook serves as supporting material for **Lecture 13 - Learning** from the lecture Grundlagen der Künstlichen Intelligenz (IN2062). This notebook is absorbed from [learning.ipynb](https://github.com/aimacode/aima-python/blob/master/learning.ipynb) and uses implementations from [learning.py](https://github.com/aimacode/aima-python/blob/master/learning.py) of the AIMA repository from the book *Artificial Intelligence: A Modern Approach*. Let's start by importing everything from the module:

In [1]:
from learning import *
from notebook import *
from utils import *

## CONTENTS

* Introduction

* Decision Tree Learner


### Introduction


You can find the code for the dataset here:

In [2]:
psource(DataSet)

### Class Attributes

* **examples**: Holds the items of the dataset. Each item is a list of values.

* **attrs**: The indexes of the features (by default in the range of [0,f), where *f* is the number of features). For example, `item[i]` returns the feature at index *i* of *item*.

* **attrnames**: An optional list with attribute names. For example, `item[s]`, where *s* is a feature name, returns the feature of name *s* in *item*.

* **target**: The attribute a learning algorithm will try to predict. By default the last attribute.

* **inputs**: This is the list of attributes without the target.

* **values**: A list of lists which holds the set of possible values for the corresponding attribute/feature. If initially `None`, it gets computed (by the function `setproblem`) from the examples.

* **distance**: The distance function used in the learner to calculate the distance between two items. By default `mean_boolean_error`.

* **name**: Name of the dataset.

* **source**: The source of the dataset (url or other). Not used in the code.

* **exclude**: A list of indexes to exclude from `inputs`. The list can include either attribute indexes (attrs) or names (attrnames).

### Class Helper Functions

These functions help modify a `DataSet` object to your needs.

* **sanitize**: Takes as input an example and returns it with non-input (target) attributes replaced by `None`. Useful for testing. Keep in mind that the example given is not itself sanitized, but instead a sanitized copy is returned.

* **classes_to_numbers**: Maps the class names of a dataset to numbers. If the class names are not given, they are computed from the dataset values. Useful for classifiers that return a numerical value instead of a string.

* **remove_examples**: Removes examples containing a given value. Useful for removing examples with missing values, or for removing classes (needed for binary classifiers).


## DECISION TREE LEARNER

### Overview

#### Decision Trees
A decision tree is a flowchart that uses a tree of decisions and their possible consequences for classification. At each non-leaf node of the tree an attribute of the input is tested, based on which a corresponding branch leading to a child-node is selected. At the leaf node the input is classified based on the class label of this leaf node. The paths from root to leaves represent classification rules based on which leaf nodes are assigned class labels.
<img src="img/decision_tree_learner_restaurant.png" alt="Drawing" style="width: 600px;"/>
#### Decision Tree Learning
Decision tree learning is the construction of a decision tree from class-labeled training data. The data is expected to be a tuple in which each record of the tuple is an attribute used for classification. The decision tree is built top-down, by choosing a variable at each step that best splits the set of items. There are different metrics for measuring the "best split". These generally measure the homogeneity of the target variable within the subsets.

#### Gini Impurity
Gini impurity of a set is the probability of a randomly chosen element to be incorrectly labeled if it was randomly labeled according to the distribution of labels in the set.

$$I_G(p) = \sum{p_i(1 - p_i)} = 1 - \sum{p_i^2}$$

We select a split which minimizes the Gini impurity in child nodes.

#### Information Gain
Information gain is based on the concept of entropy from information theory. Entropy is defined as:

$$B(p) = -\sum{p_i \log_2{p_i}}$$

Information Gain is difference between entropy of the parent and weighted sum of entropy of children. The feature used for splitting is the one which provides the most information gain.

#### Pseudocode

You can view the pseudocode by running the cell below:

In [3]:
pseudocode("Decision Tree Learning")

### AIMA3e
__function__ DECISION-TREE-LEARNING(_examples_, _attributes_, _parent\_examples_) __returns__ a tree  
&emsp;__if__ _examples_ is empty __then return__ PLURALITY\-VALUE(_parent\_examples_)  
&emsp;__else if__ all _examples_ have the same classification __then return__ the classification  
&emsp;__else if__ _attributes_ is empty __then return__ PLURALITY\-VALUE(_examples_)  
&emsp;__else__  
&emsp;&emsp;&emsp;_A_ &larr; argmax<sub>_a_ &isin; _attributes_</sub> IMPORTANCE(_a_, _examples_)  
&emsp;&emsp;&emsp;_tree_ &larr; a new decision tree with root test _A_  
&emsp;&emsp;&emsp;__for each__ value _v<sub>k</sub>_ of _A_ __do__  
&emsp;&emsp;&emsp;&emsp;&emsp;_exs_ &larr; \{ _e_ : _e_ &isin; _examples_ __and__ _e_._A_ = _v<sub>k</sub>_ \}  
&emsp;&emsp;&emsp;&emsp;&emsp;_subtree_ &larr; DECISION-TREE-LEARNING(_exs_, _attributes_ &minus; _A_, _examples_)  
&emsp;&emsp;&emsp;&emsp;&emsp;add a branch to _tree_ with label \(_A_ = _v<sub>k</sub>_\) and subtree _subtree_  
&emsp;&emsp;&emsp;__return__ _tree_  

---
__Figure ??__ The decision\-tree learning algorithm. The function  IMPORTANCE is described in Section __??__. The function PLURALITY\-VALUE selects the most common output value among a set of examples, breaking ties randomly.

### Implementation
The nodes of the tree constructed by our learning algorithm are stored using either `DecisionFork` or `DecisionLeaf` based on whether they are a parent node or a leaf node respectively.

In [4]:
psource(DecisionFork)

`DecisionFork` holds the attribute, which is tested at that node, and a dict of branches. The branches store the child nodes, one for each of the attribute's values. Calling an object of this class as a function with input tuple as an argument returns the next node in the classification path based on the result of the attribute test.

In [5]:
psource(DecisionLeaf)

The leaf node stores the class label in `result`. All input tuples' classification paths end on a `DecisionLeaf` whose `result` attribute decide their class.

In [6]:
psource(DecisionTreeLearner)

The implementation of `DecisionTreeLearner` provided in [learning.py](https://github.com/aimacode/aima-python/blob/master/learning.py) uses information gain as the metric for selecting which attribute to test for splitting. The function builds the tree top-down in a recursive manner. Based on the input it makes one of the four choices:
<ol>
<li>If the input at the current step has no training data we return the mode of classes of input data received in the parent step (previous level of recursion).</li>
<li>If all values in training data belong to the same class it returns a `DecisionLeaf` whose class label is the class which all the data belongs to.</li>
<li>If the data has no attributes that can be tested we return the class with highest plurality value in the training data.</li>
<li>We choose the attribute which gives the highest amount of entropy gain and return a `DecisionFork` which splits based on this attribute. Each branch recursively calls `decision_tree_learning` to construct the sub-tree.</li>
</ol>

### Example

We will now use the Decision Tree Learner to learn a decision tree for the example from **lecture 13, slide 13**.
First, create the dataset.

In [7]:
attrnames = ['Alt', 'Bar', 'Fri', 'Hun', 'Pat', 'Price', 'Rain', 'Res', 'Type', 'Est', 'WillWait']
examples = [
            ['T', 'F', 'F', 'T', 'Some', '$$$', 'F', 'T', 'French',  '0-10',  'T'],
            ['T', 'F', 'F', 'T', 'Full', '$',   'F', 'F', 'Thai',    '30-60', 'F'], 
            ['F', 'T', 'F', 'F', 'Some', '$',   'F', 'F', 'Burger',  '0-10',  'T'], 
            ['T', 'F', 'T', 'T', 'Full', '$',   'F', 'F', 'Thai',    '10-30', 'T'], 
            ['T', 'F', 'T', 'F', 'Full', '$$$', 'F', 'T', 'French',  '>60',   'F'], 
            ['F', 'T', 'F', 'T', 'Some', '$$',  'T', 'T', 'Italian', '0-10',  'T'], 
            ['F', 'T', 'F', 'F', 'None', '$',   'T', 'F', 'Burger',  '0-10',  'F'], 
            ['F', 'F', 'F', 'T', 'Some', '$$',  'T', 'T', 'Thai',    '0-10',  'T'], 
            ['F', 'T', 'T', 'F', 'Full', '$',   'T', 'F', 'Burger',  '>60',   'F'], 
            ['T', 'T', 'T', 'T', 'Full', '$$$', 'F', 'T', 'Italian', '10-30', 'F'], 
            ['F', 'F', 'F', 'F', 'None', '$',   'F', 'F', 'Thai',    '0-10',  'F'], 
            ['T', 'T', 'T', 'T', 'Full', '$',   'F', 'F', 'Burger',  '30-60', 'T']
           ]

lecture_dataset = DataSet(examples=examples, attr_names=attrnames, target='WillWait')
lecture_dataset

<DataSet(): 12 examples, 11 attributes>

In [8]:
DTL = DecisionTreeLearner(lecture_dataset)

Take a look at the tree.

In [9]:
DTL

DecisionFork(4, 'Pat', {'Some': 'T', 'None': 'F', 'Full': DecisionFork(3, 'Hun', {'F': 'F', 'T': DecisionFork(8, 'Type', {'French': 'T', 'Thai': DecisionFork(9, 'Est', {'10-30': 'T', '0-10': 'F', '>60': 'T', '30-60': 'F'}), 'Burger': 'T', 'Italian': 'F'})})})

Display the tree properly.

In [10]:
DTL.display()

Test Pat
 Pat = Some ==> RESULT = T
 Pat = None ==> RESULT = F
 Pat = Full ==> Test Hun
 Hun = F ==> RESULT = F
 Hun = T ==> Test Type
 Type = French ==> RESULT = T
 Type = Thai ==> Test Est
 Est = 10-30 ==> RESULT = T
 Est = 0-10 ==> RESULT = F
 Est = >60 ==> RESULT = T
 Est = 30-60 ==> RESULT = F
 Type = Burger ==> RESULT = T
 Type = Italian ==> RESULT = F


<div class="alert alert-info">
    <h3>Note</h3>
    <p>The calculated decision tree could be different from the one showed in the lecture slide 19. The reason is that after spliting the samples with Patrons, the information gain for the five attributes: Hun, Price, Res, Type, and Est are equal. Therefore, the algorithm randomly chooses one of the five attribute and split the samples further. We illustrate this by calculating the information gain values in the following.</p> 
</div>

<img src="img/decision_tree_restaurant_result.png" alt="Drawing" style="width: 500px;"/>

We adapt the functions from learning.py first.

In [11]:
def split_by(attr, examples, values):
        """Return a list of (val, examples) pairs for each val of attr."""
        return [(v, [e for e in examples if e[attr] == v])
                for v in values[attr]]

def count(attr, val, examples):
    """Count the number of examples that have example[attr] = val."""
    return sum(e[attr] == val for e in examples)

def information_gain(attr, examples, values, target):
    """Return the expected reduction in entropy from splitting by attr."""
    def I(examples):
        return information_content([count(target, v, examples)
                                    for v in values[target]])
    N = len(examples)
    remainder = sum((len(examples_i)/N) * I(examples_i)
                    for (v, examples_i) in split_by(attr, examples, values))
    return I(examples) - remainder

We get three child datasets after spliting the samples with attribute Patrons. All the samples of the child dataset for Pat="None" and for Pat="Some" have the same value for the target. So we only need to split the tree further for Pat="Full". We define the child dataset first:

In [12]:
attrnames = ['Alt', 'Bar', 'Fri', 'Hun', 'Price', 'Rain', 'Res', 'Type', 'Est', 'WillWait']
examples = [
            ['T', 'F', 'F', 'T', '$',   'F', 'F', 'Thai',    '30-60', 'F'], 
            ['T', 'F', 'T', 'T', '$',   'F', 'F', 'Thai',    '10-30', 'T'], 
            ['T', 'F', 'T', 'F', '$$$', 'F', 'T', 'French',  '>60',   'F'],
            ['F', 'T', 'T', 'F', '$',   'T', 'F', 'Burger',  '>60',   'F'], 
            ['T', 'T', 'T', 'T', '$$$', 'F', 'T', 'Italian', '10-30', 'F'],
            ['T', 'T', 'T', 'T', '$',   'F', 'F', 'Burger',  '30-60', 'T']
           ]
child_dataset = DataSet(examples=examples, attr_names=attrnames, target='WillWait')
child_dataset

<DataSet(): 6 examples, 10 attributes>

In [13]:
for attr in child_dataset.inputs:
    info_gain = information_gain(attr, child_dataset.examples, 
                                 child_dataset.values, child_dataset.target)
    print("Information gain of attribute {} is {}".format(child_dataset.attr_names[attr], info_gain))

Information gain of attribute Alt is 0.109170338675599
Information gain of attribute Bar is 0.0
Information gain of attribute Fri is 0.109170338675599
Information gain of attribute Hun is 0.2516291673878229
Information gain of attribute Price is 0.2516291673878229
Information gain of attribute Rain is 0.109170338675599
Information gain of attribute Res is 0.2516291673878229
Information gain of attribute Type is 0.2516291673878229
Information gain of attribute Est is 0.2516291673878229


We can see that the information gain of Hun, Price, Res, Type, and Est are all 0.2516. So it does not matter along which of the five attributes to split further. 

Now we use the learned decision tree to classify a sample with 
    'Alt'= 'F', 'Bar'='F', 'Fri'='F', 'Hun='F', 'Pat'='None', 'Price'='$', 'Rain'='F', 'Res'='F', 'Type'='Thai', 'Est'='0-10'.

In [14]:
print(DTL(['F', 'F', 'F', 'F','None', '$', 'F', 'F', 'Thai', '0-10']))

F


As expected, the Decision Tree learner classifies the sample as "F".